Have the build phase reparse the BuildConfigurationFrom the environment.

mattmoor opened the pull request:
* A short explanation of the proposed change:

Rebuild the BuildConfiguration from the environment to allow build phases prior to go-build in the order to influence the way go-build executes.  A canonical example of this would be a function framework injecting a `package main` entrypoint around a function that should be invoked when requests are received.

Fixes: https://github.com/paketo-buildpacks/go-build/issues/91

* An explanation of the use cases your change enables:

This allows downstream buildpacks to synthesize an entrypoint during their build phase and direct go-build to build that target by setting `BP_GO_TARGETS`.

Please confirm the following:
* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [ ] I have added an integration test, if necessary.

ryanmoran commented:
Thanks for putting in the time to PR this. The diff looks reasonable but, it also makes sense to remove some code from the detect along with this change.

We can remove the code that populates the metadata as its no longer used during build: https://github.com/paketo-buildpacks/go-build/blob/3f53b865624812dfc5c04e453a69fde8f2451dce/detect.go#L19-L29

Then, as there is no metadata to convey, we can remove the provision and requirement of the `go-build` entry:
https://github.com/paketo-buildpacks/go-build/blob/3f53b865624812dfc5c04e453a69fde8f2451dce/detect.go#L33
https://github.com/paketo-buildpacks/go-build/blob/3f53b865624812dfc5c04e453a69fde8f2451dce/detect.go#L41-L44

Finally, we need to keep the invocation of `parser.Parse` function as it will error in a case where we want the buildpack to fail detection, though its non-error return value can be discarded: https://github.com/paketo-buildpacks/go-build/blob/3f53b865624812dfc5c04e453a69fde8f2451dce/detect.go#L14-L17

With these changes, I'd be happy to test this out again and merge.

mattmoor commented:
Cool, I rebased this, since github was complaining.  The changes are in a new commit.  I removed the bits you highlighted above, and (given that we are dropping the plan), I remove the plan bits from `build*.go` as well.
